### ISSUE TITLE
Nonliner conjugate gradient optimizer + Hager-Zhang line search

### ISSUE THREAD

**Issue by piyueh** [pytorch/pytorch#80553]
### 🚀 The feature, motivation and pitch

The nonlinear conjugate gradient (CG) method is a good alternative to the L-BFGS optimizer. Features of nonlinear CG:

1. It theoretically converges faster than gradient-descent methods (including most derived variants), and the memory footprint is just slightly higher than that of gradient-descent methods.
2. It is theoretically able to converge to machine precision, which may be important for regression-type problems when using single-precision floats.
3. Though it converges slower than L-BFGS does, it needs much less memory.
4. The specific CG method from Hager and Zhang automatically falls back to the gradient descent method if the conjugate direction does not perform better.
5. No need for configuring the learning rate. The step size (i.e., the learning rate) on the conjugate/gradient direction is determined by an inexact line search.
6. Hager and Zhang also proposed a companion inexact line-search algorithm for the proposed CG method.

I already have a working implementation of Hager and Zhang's nonlinear CG as a derived `torch.optim.Optimizer` class, including the inexact line search algorithm. I have been using this implementation for my own research. However, in terms of code tests, I only have two simple cases that can be used for code tests. But at least if this feature request is approved, we don't have to start from scratch.

This is directly related to #1359.

**Reference**
[1] W. W. Hager and H. Zhang, “Algorithm 851: CG_DESCENT, a conjugate gradient method with guaranteed descent,” ACM Trans. Math. Softw., vol. 32, no. 1, pp. 113–137, Mar. 2006, doi: 10.1145/1132973.1132979.

### Additional context

There are some issues mentioning implementing CG, like #53441 and #17902. However, I think they are more about linear CG, which is a special case of nonlinear CG. And I feel like linear and nonlinear CG solvers usually are implemented separately. So I don't think this feature request will resolve the issues requesting linear CG.

**Comment by BraveDrXuTF** [comment]
Has the CG been implemented?
